### ISSUE TITLE
Samba integration

### ISSUE THREAD

**Issue by alsvartr** [nitnelave/lldap#599]
Hello, can lldap be used for samba authentication?
Trying to integrate it with samba gives me following errors:
>  ../../source3/passdb/pdb_ldap_util.c:313(smbldap_search_domain_info)
smbldap_search_domain_info: Adding domain info for LDAP_SERVER failed with NT_STATUS_UNSUCCESSFUL
../../source3/passdb/pdb_ldap.c:6751(pdb_ldapsam_init_common)
pdb_init_ldapsam: WARNING: Could not get domain info, nor add one to the domain. We cannot work reliably without it.
../../source3/passdb/pdb_interface.c:179(make_pdb_method_name)
pdb backend ldapsam:ldap://ldap_server:3890 did not correctly init (error was NT_STATUS_CANT_ACCESS_DOMAIN_INFO)

I assume that lldap just missing some schemas to work with samba?

**Comment by nitnelave** [comment]
I haven't looked into it, but I'd wager that sambda requires some non-lldap-default fields, which would be blocked by https://github.com/lldap/lldap/issues/67
But it's being (slowly) worked on!

That said, a cursory search mostly brings up sambda acting as a ldap server for linux, rather than the opposite. How are you trying to set it up?

**Comment by alsvartr** [comment]
> That said, a cursory search mostly brings up sambda acting as a ldap server for linux, rather than the opposite. How are you trying to set it up?

I'm trying to set simple auth via LDAP. Like this (example from 'Samba with OpenLDAP' tutorials):
> passdb backend = ldapsam:ldap://ldap_server:3890
    ldap suffix = dc=domain,dc=tld
    ldap admin dn = uid=bind_ro,ou=people,dc=domain,dc=tld
    ldap ssl = no
    ldap user suffix = ou=people
    ldap group suffix = ou=groups

**Comment by pcmediapear** [comment]
+1 for samba integration
same issue on qnap nas:
> [2023/07/02 21:32:57.590384,  0] ../../source3/passdb/pdb_ldap_util.c:314(smbldap_search_domain_info)
>   smbldap_search_domain_info: Adding domain info for DOMAIN.NAME failed with NT_STATUS_UNSUCCESSFUL
> [2023/07/02 21:32:57.590479,  0] ../../source3/passdb/pdb_ldap.c:6756(pdb_ldapsam_init_common)
>   pdb_init_ldapsam: WARNING: Could not get domain info, nor add one to the domain. We cannot work reliably without it.
> [2023/07/02 21:32:57.590512,  0] ../../source3/passdb/pdb_interface.c:186(make_pdb_method_name)
>   pdb backend ldapsam:ldaps://LDAP.SERVER did not correctly init (error was NT_STATUS_CANT_ACCESS_DOMAIN_INFO)

**Comment by ipsi** [comment]
I'd also like to see this. I'd like to use something simpler than OpenLDAP but my Synology NAS needs some amount of Samba stuff - I _think_ just the password control portions of it, but not 100% sure.

**Comment by nitnelave** [comment]
@ipsi FWIW, I'm not sure that LLDAP will ever be compatible with Synology, even with Samba. Last time I looked, Synology was requesting the hashed password to be able to check the login locally. That's something we just can't do in LLDAP, by design (we don't store the hashed password, but instead we use a zero-knowledge protocol to verify the password)

**Comment by ipsi** [comment]
I think that topic is more complex (and completely unrelated to Samba). Briefly, though, I believe Synology uses SSSD to communicate with the LDAP server. SSSD can _optionally_ cache passwords, and can also be used to _change_ the password. If you don't do either, it shouldn't matter if it doesn't return the `userPassword` attribute.

Once it's possible to get Samba attributes in, I might have a poke and see if I'm right about that.

**Comment by nitnelave** [comment]
Alright. Feel free to follow the progress at https://github.com/lldap/lldap/issues/67
If you're feeling exceptionally motivated, you can already create the fields and set the values directly in the database, and they'll be returned over LDAP, so you can directly check your assumptions. But I totally understand if you want to wait until I actually implement field creation/setting in the web UI :)

**Comment by eudaldgr** [comment]
> I think that topic is more complex (and completely unrelated to Samba). Briefly, though, I believe Synology uses SSSD to communicate with the LDAP server. SSSD can _optionally_ cache passwords, and can also be used to _change_ the password. If you don't do either, it shouldn't matter if it doesn't return the `userPassword` attribute.
> 
> Once it's possible to get Samba attributes in, I might have a poke and see if I'm right about that.

Have you already tried something or have you had success with this? I am interested in this very thing.

**Comment by nitnelave** [comment]
By the way, #67  should be ready now, you can create custom attributes with https://github.com/Zepmann/lldap-cli

Who wants to give samba a try?

**Comment by johnmmcgee** [comment]
I would love to see this integration working.

**Comment by nitnelave** [comment]
@johnmmcgee we just need a volunteer to figure out the configuration. Want to give it a try?

**Comment by johnmmcgee** [comment]
sure.  how would one create these fields?  My ldap experience is not that great, so any direction would be welcome.

**Comment by nitnelave** [comment]
Usually that requires looking at the docs of the service to see what fields they expect (they sometimes provide an ldif file that outlines the schema they expect), and/or reverse engineering their expectations from the services' debug logs and LLDAP verbose mode.

Then using lldap-cli, you can create the user/group attributes required.

Feel free to head over to LLDAP's discord server if you have any questions.

**Comment by nitnelave** [comment]
Now that we have full UI support for custom attributes, does anyone want to give this a try?

**Comment by nitnelave** [comment]
Can you try creating that attribute? That should be easy with the web UI (though it has to be populated for each user)

**Comment by pcmediapear** [comment]
sorry, it is not an user attribute
if I good understand it is an separate entry in root of ldap tree
https://wiki.samba.org/index.php/3.0:_Initialization_LDAP_Database

**Comment by MassiPi** [comment]
hello,
is this what prevents openmediavault to be able to use lldap?
i like a lot the starting point of this project, but being able to work with samba seems a big point for A LOT of homelabs, isn't it? (and also integration with OMV seems a big point..)
thanks

**Comment by nitnelave** [comment]
> is this what prevents openmediavault to be able to use lldap?

I don't know OMV very well, but it seems to be Linux based, and using the system users. Rather than Samba integration (more adapted for windows), you're probably looking for PAM integration to get the LLDAP users in the system. There is a guide for this (linked from the readme).

**Comment by MassiPi** [comment]
yes i am very interested in this. I mean, OMV really uses /etc/group /etc/passwd and so on, afaik configuring pam and the other services is just going to add users to those files, isn't it?
ok i need to try to break something :) 
OMV would be a great addon since i see you already have NextCloud as working..

**Comment by thomaschampagne** [comment]
Up vote 👍 

Looking for this feature too 😃 

Thanks for your work !

**Comment by nitnelave** [comment]
Sorry to disappoint, but I don't foresee Samba integration anytime soon, if ever. I had a look, and Samba is a beast, expecting you to define machines, roles and so on as part of the LDAP server.

It would be very complicated to integrate with LLDAP, and honestly probably not worth it.

**Comment by MassiPi** [comment]
> Samba is a beast

yes it's unplayable. I'm trying to get OMV to authenticate over ldap, and also with (way) heavier and complex solution like OpenDJ where getting sssd to work is trivial, Samba is still a complete mess.
I'm sure there will be a proper reason, but it's still unbelievebly hard to sort out lol

**Comment by thomaschampagne** [comment]
> Sorry to disappoint, but I don't foresee Samba integration anytime soon, if ever. I had a look, and Samba is a beast, expecting you to define machines, roles and so on as part of the LDAP server.
> 
> It would be very complicated to integrate with LLDAP, and honestly probably not worth it.

I understand. I will give a try through sssd (https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/7/html/windows_integration_guide/smb-sssd#SMB-SSSD)

Thanks again !

**Comment by divStar** [comment]
Please excuse me for asking this (I have read through this issue to the best of my abilities), but I couldn't figure out whether it is possible to have Samba use LLDAP for authentication for file shares - I personally do not use Samba AD or anything like that. Maybe adding the custom attributes would help? But the special object classes Samba expects would probably still be absent - are they needed for the file shares though?

**Comment by terziman** [comment]
For futures reference, I managed to have a partial integration of LLDAP in samba.

What works:
- user names from lldap
- correct uid & gid from lldap
- shares permissions from groups defined in lldap

What do not work (and probably never will):
- login with lldap password

Indeed, the only solution I found for now is to keep a local db in samba for the hashed passwords and let samba authenticate users on it's own. I do not need however to create system users and groups anymore.

How it work: I configured samba to lookup user and group info from PAM and sssd. I basically used the example config in the doc for sssd wish the added following options in sssd.conf:
> ldap_schema = rfc2307bis
> ldap_user_uuid = dn
> ldap_group_uuid = dn
> override_homedir = /home/%u
> default_shell = /sbin/nologin

And I added custom integer attributes in lldap only for uidNumber, gidNumber (users) and gidNumber (groups) as I was not interested by the other attributes.

On samba side, I only had to add the following option in smb.conf:
> obey pam restrictions = yes

And added the users passwords with smbpasswd as usual.

The tricky part was that I actually run samba inside a docker container. This way, sssd does not give access to the host in any way but I got quite some trouble to get it working. I first started with nslcd instead of sssd, but I only managed to get it find the users and groups from lldap, but not find which user belongs to which group. After a lot of tinkering, I finally switched to sssd, but then everything stopped working. I was using Alpine as a base image, so I switched to an ubuntu base image and sssd finally started to work properly. I need the following packages: sssd, libpam-sss and libnss-sss

/etc/nsswitch.conf must be edited to add:
> passwd: files sssd
> group: files sssd
> shadow: files sssd

And pam configurations updated with:
> pam-auth-update --enable sssd

To test if lldap users and groups were properly retrieved, I used:
> getent passwd
> getent passwd <ldap_user>
> getent group

And I verified that lldap groups were properly assigned to lldap users, as well as the users uid and gid.

I hope it can help others spend less time than I did to figure it all out. And if anybody has any idea to make samba login work with lldap credentials, I'm interested.

(shell login with lldap users works inside the container, but I did not managed to forward plain text password from samba to pam,  and I believe it is deprecated anyway and blocked)

**Comment by terziman** [comment]
To work around the samba password limitation, I believe one way would be to have some way to synchronize lldap and samba passwords.

The easiest way I see would be to add some kind of notifications using API hooks in lldap to call custom scripts for some some operations:
- add user (parameter: uid and password)
- remove user (parameter: uid)
- update user password (parameter: uid and password)

I don't know if it could be feasible or if it makes sense for lldap architecture in general.

**Comment by nitnelave** [comment]
That wouldn't work. LLDAP is designed to not store _or receive_ user passwords, using a zero knowledge proof instead. We literally _can't_ send the password to samba

**Comment by terziman** [comment]
What about from the frontend?

**Comment by nitnelave** [comment]
While theoretically possible, this would require a double implementation (backend to samba and frontend to samba) and not something I'd consider.

The easiest would be to fork LLDAP and remove the OPAQUE part (zero-knowledge proof). It should be fairly simple since we also support simple auth
